### PR TITLE
Feat(mlx): Expose rmsprop/adamax/adagrad/adadelta optimizers in MLXTrainer

### PR DESCRIPTION
**Pull request by BardiaKoopah**
## Summary

Adds `rmsprop`, `adamax`, `adagrad`, and `adadelta` to the optimizers
supported by `MLXTrainer`, completing Item-6 of the MLX feature audit.
Previously these names hit the unsupported-optimizer path and were
silently coerced to AdamW, so recipes tuned on them diverged on the MLX
backend.

## Changes

- Extend `SUPPORTED_MLX_OPTIMIZERS` (trainer.py:52) with the four names.
- Add four `elif` branches in `_build_optimizer`, matching the existing
  adamw/muon/lion pattern.

## Notes / deliberate choices

- **Adamax** threads `**adam_kwargs` (carries `betas`); the other three
  take `learning_rate` only, since their MLX signatures don't accept
  `betas`. Verified against `inspect.signature` on the pinned mlx build.
- Weight decay routes through the existing `_apply_manual_weight_decay`
  path via `self._manual_weight_decay`, matching AdamW/Muon/Lion
  (decoupled) semantics rather than SGD's coupled path.
- Class casing matches `mlx.optimizers` exactly: `RMSprop`, `Adamax`,
  `Adagrad`, `AdaDelta`.
- Grad-norm reporting is skipped for rmsprop/adagrad/adadelta by existing
  trainer logic (the "skipping grad norm reporting" notice) — expected,
  not introduced here.

## Testing

Smoke-tested all four on a 5-step gemma-3-4b-it VLM LoRA run
(`load_in_4bit`, CCE). Each constructs, trains through 5 steps without
error, and saves adapters. `_resolved_optimizer_name` matches the
requested optimizer in every case.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/trainer.py [hunk 1]
```diff
@@ -1360,6 +1360,18 @@ def _build_optimizer(self, total_steps):
         elif opt_name == "lion":
             self._manual_weight_decay = float(wd or 0.0)
             optimizer = optim.Lion(learning_rate=initial_lr, weight_decay=0.0)
+        elif opt_name == "rmsprop":
+            self._manual_weight_decay = float(wd or 0.0)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: Fixed in 754c07d: rmsprop, adamax, adagrad and adadelta now use coupled L2 weight decay (grad += wd*param), matching torch.optim and the existing SGD branch, instead of AdamW-style decoupled decay. This matters at the default weight_decay=0.001, where decoupled decay changed the trained result versus the requested optimizer. The three Adamax/Adagrad diagnostics-and-defaults notes are below the correctness bar (reporting-only grad_norm under non-default max_grad_norm<=0; a bias-correction warmup transient with no MLX kwarg to set it; and a larger, more stable Adagrad eps), so left as-is.

File: unsloth_zoo/mlx/trainer.py [hunk 2]
```diff
@@ -1360,6 +1360,24 @@ def _build_optimizer(self, total_steps):
...
             optimizer = optim.Lion(learning_rate=initial_lr, weight_decay=0.0)
+        elif opt_name == "rmsprop":
+            # HF/PyTorch RMSprop, Adamax, Adagrad, and Adadelta fold weight
+            # decay into the gradient as coupled L2 (grad += wd * param),
+            # unlike AdamW's decoupled shrink. Use the same coupled path as
+            # SGD so the adaptive denominator and optimizer state see the
+            # decay term and the trained result matches the requested
+            # optimizer.
+            self._coupled_weight_decay = float(wd or 0.0)
+            optimizer = optim.RMSprop(learning_rate=initial_lr)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py:3572]: The MLX trainer is a simplified backend that does not consume optim_args for any optimizer: adamw/adam/sgd/muon/lion also ignore momentum/alpha/centered/lr_decay and read only the named config fields (adam_beta1/adam_beta2, weight_decay, learning_rate). rmsprop/adamax/adagrad/adadelta are consistent with that existing behavior, so this is not a regression from adding them. Threading arbitrary HF optim_args through the MLX backend (or failing fast on unsupported ones) is a separate, cross-optimizer feature.

File: unsloth_zoo/mlx/trainer.py [hunk 3]
```diff
@@ -1360,6 +1360,24 @@ def _build_optimizer(self, total_steps):
...
+        elif opt_name == "adamax":
+            self._coupled_weight_decay = float(wd or 0.0)
+            optimizer = optim.Adamax(learning_rate=initial_lr, **adam_kwargs)
+        elif opt_name == "adagrad":
+            self._coupled_weight_decay = float(wd or 0.0)
+            optimizer = optim.Adagrad(learning_rate=initial_lr)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: Same as the RMSprop note: the MLX backend does not apply optim_args (here lr_decay) for any optimizer, so adagrad matches the existing adamw/sgd/etc. behavior rather than regressing. A general optim_args passthrough is out of scope for this PR.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/trainer.py [hunk 4]
```diff
@@ -1360,6 +1360,24 @@ def _build_optimizer(self, total_steps):
...
+            # decay into the gradient as coupled L2 (grad += wd * param),
+            # unlike AdamW's decoupled shrink. Use the same coupled path as
+            # SGD so the adaptive denominator and optimizer state see the
+            # decay term and the trained result matches the requested
+            # optimizer.
+            self._coupled_weight_decay = float(wd or 0.0)
+            optimizer = optim.RMSprop(learning_rate=initial_lr)
+        elif opt_name == "adamax":
+            self._coupled_weight_decay = float(wd or 0.0)
+            optimizer = optim.Adamax(learning_rate=initial_lr, **adam_kwargs)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: MLX's Adamax is a correct implementation. The missing first-moment bias correction only affects the first few steps (1/(1 - beta1^t) approaches 1 as t grows, so it washes out within warmup), and MLX's Adamax exposes no bias_correction option, so matching PyTorch exactly would mean reimplementing the optimizer. The optimizer trains correctly, so this is an HF parity nuance rather than a correctness bug.

[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: Good catch, fixed. Adamax exposed betas so it passed the Adam-family gate, but its v state is an infinity norm (v_t = max(beta2*v_{t-1}, |g_t|)), not the exponential second moment the recovery formula assumes, so the reported grad_norm was wildly scaled. It now routes through a shared _mlx_optimizer_state_norm_supported helper that excludes Adamax, so it is skipped like the other non-Adam optimizers. Added a unit test.

File: unsloth_zoo/mlx/trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: MLX's Adagrad default eps=1e-8 is a valid and slightly more stable choice. The 1e-8 vs 1e-10 difference only matters for very small or sparse gradients and does not make the optimizer incorrect, so this is an HF parity nuance rather than a correctness bug. Keeping MLX's default.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/trainer.py [hunk 5]
```diff
@@ -1360,6 +1377,24 @@ def _build_optimizer(self, total_steps):
...
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: As on the earlier rounds: MLX's Adamax is a correct implementation and exposes no bias_correction option, so matching PyTorch's first-moment correction would require reimplementing the optimizer. The effect is a transient early-step scaling (1/(1 - beta1^t) approaches 1 within a few steps, well inside warmup) and the optimizer trains correctly, so this remains an HF parity nuance rather than a correctness bug.

File: unsloth_zoo/mlx/trainer.py [hunk 6]
```diff
@@ -1360,6 +1377,24 @@ def _build_optimizer(self, total_steps):
...
+        elif opt_name == "adagrad":
+            self._coupled_weight_decay = float(wd or 0.0)
+            optimizer = optim.Adagrad(learning_rate=initial_lr)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: As on the earlier rounds: MLX's Adagrad default eps=1e-8 is a valid and slightly more numerically stable choice. The 1e-8 vs 1e-10 difference only matters for pathologically small or sparse gradients and does not make the optimizer incorrect, so this remains an HF parity nuance rather than a correctness bug. Keeping MLX's default.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: As noted across the earlier rounds: MLX's Adamax exposes no bias_correction option, and the missing first-moment correction only scales the first few steps (1/(1 - beta1^t) approaches 1 within a few steps, well inside warmup). The optimizer trains correctly, so this remains an HF parity nuance rather than a correctness bug.

File: unsloth_zoo/mlx/trainer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: As noted previously: MLX's Adagrad default eps=1e-8 is a valid and slightly more numerically stable choice. The 1e-8 vs 1e-10 difference only matters for pathologically small or sparse gradients and does not make the optimizer incorrect, so this remains a parity nuance. Keeping MLX's default.

File: unsloth_zoo/mlx/trainer.py [hunk 7]
```diff
@@ -52,7 +52,24 @@
 from mlx.utils import tree_flatten, tree_map, tree_reduce, tree_unflatten
 
 _PAD_MULTIPLE = 32
-SUPPORTED_MLX_OPTIMIZERS = ("adafactor", "adamw", "adam", "sgd", "muon", "lion")
+SUPPORTED_MLX_OPTIMIZERS = ("adafactor", "adamw", "adam", "sgd", "muon", "lion", "rmsprop", "adamax", "adagrad", "adadelta")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: The rmsprop_bnb / rmsprop_8bit / rmsprop_32bit variants are bitsandbytes-specific optimizers that MLX cannot provide. The trainer already fails with a clear unsupported-optimizer error that lists the supported set, which is the deliberate-error resolution you note. Coercing them to plain rmsprop would silently drop the user's requested 8-bit optimizer state, so failing clearly is the safer behavior here. Keeping it.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: As covered across the prior rounds: MLX's Adamax exposes no bias_correction option, and the missing first-moment correction only scales the first few steps (within warmup); the optimizer trains correctly, so this remains an HF parity nuance rather than a correctness bug.

File: unsloth_zoo/mlx/trainer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py]: As covered previously: MLX's Adagrad default eps=1e-8 is a valid and slightly more numerically stable choice; the 1e-8 vs 1e-10 difference only matters for pathologically small or sparse gradients and does not make the optimizer incorrect. Keeping MLX's default.

**Comment by danielhanchen** [comment]
## Validation on real Apple Silicon (Metal)

Ran this branch end to end on a macOS Apple Silicon runner with real MLX on a real Metal GPU (no torch shim) via staging CI (head `ab252d0`). Stack: mlx 0.32.0, mlx-metal 0.32.0, mlx-lm 0.31.3, transformers 5.13.0.

### Real-MLX feature validation

| Check | Result |
|---|---|
| real MLX / Metal available | PASS |
| import unsloth_zoo.mlx | PASS |
| pr819_new_optimizers (API) | PASS |
| pr819_optimizer_training (real training) | PASS |
| shared LoRA SFT smoke | PASS |
| Total | 5 pass, 8 skip, 0 fail |

### Real optimizer training smokes on Metal

Each new optimizer was run through a short real-training smoke (same seed, same 3-step LoRA SFT config). All produce distinct, finite updates, confirming the optimizer actually steps on the real MLX path rather than being a no-op:

| Optimizer run | Step-3 loss | Peak mem | Wall |
|---|---|---|---|
| run 1 | 3.1050 | 0.25 GB | 2.0s |
| run 2 | 3.8606 | 0.36 GB | 1.7s |
| run 3 | 3.8492 | 0.44 GB | 1.8s |
| run 4 | 3.8786 | 0.55 GB | 2.8s |

Plus the shared 6-step LoRA SFT smoke: avg loss 3.5153, 235 tokens, wall 12.0s, peak 0.17 GB, loss 3.5637 to 2.9962.

**Comment by BardiaKoopah** [comment]
@codex review

**Comment by BardiaKoopah** [comment]
@danielhanchen please merge

**Comment by BardiaKoopah** [comment]
@codex review

**Comment by BardiaKoopah** [comment]
@codex review

**Comment by BardiaKoopah** [comment]
@codex review

**Comment by BardiaKoopah** [comment]
@chatgpt-codex-connector the Adamax bias correction one is right, and I'd previously talked myself out of it on the grounds that HF has no `optim="adamax"` member — which doesn't answer what someone asking for Adamax actually expects. Fixed in `6dddd3f`. MLX's Adamax takes no `bias_correction` flag and overrides `Adam.apply_single` with the uncorrected update, so it needed a subclass rather than a kwarg. Measured against `torch.optim.Adamax` on this M2 Pro with mlx 0.31.2, stock MLX moves the weight 0.100, 0.139, 0.176, 0.212, 0.245 times as far over the first five steps; `_BiasCorrectedAdamax` tracks torch to 1.5e-8 max abs after five. Dropped the test that documented the old decision.

The grad-norm one is stale — `_grad_norm_from_optimizer_state` doesn't exist anymore. Main's #922 replaced the optimizer-state estimate with a real pre-clip fp32 reduction over the gradient (`_global_grad_norm_fp32`, `trainer.py:891`) and the merge left the old helper reachable only from its own test, so I removed it in `8831b3f`. Nothing reads `optimizer.betas` to decide how `grad_norm` is reported now; the gate is `report_grad_norm` / `max_grad_norm > 0`, not the optimizer family.

Not taking the RMSprop `optim_args` one though. HF Trainer doesn't honour them either for `optim="rmsprop"`: the RMSPROP branch is just `optimizer_cls = torch.optim.RMSprop`, and `optim_args` is merged into the optimizer kwargs only by the GaLore/Apollo/GrokAdamW branches. On transformers 4.57.6, `Trainer.get_optimizer_cls_and_kwargs` with `optim="rmsprop"` and `optim_args="momentum=0.9,alpha=0.95,centered=True"` returns `(torch.optim.RMSprop, {'lr': 0.0003})`, and the resulting param group reads momentum 0, alpha 0.99, centered False — those tuned values are dropped on the torch backend too. MLX RMSprop already defaults to alpha=0.99 and eps=1e-8, same as torch, so the recipe matches as built and parsing `optim_args` here would make the MLX path diverge from the torch Trainer rather than converge with it. Pinned that upstream behaviour in `771c535` so a transformers that does start forwarding them fails the test instead of drifting quietly. momentum and centered aren't expressible in MLX RMSprop anyway, so honouring them isn't on the table without writing our own.

**Comment by BardiaKoopah** [comment]
@codex review

**Comment by danielhanchen** [comment]
Confirmed these four names still hit the unsupported-optimizer path in unsloth_zoo/mlx/trainer.py and that the branches route weight decay through the existing coupled path like the SGD branch, so the requested change is in. Will get this reviewed.

**Comment by danielhanchen** [comment]
Checked that unsloth_zoo/mlx/trainer.py:58 still stops at adamw/adam/sgd/muon/lion plus the 8bit pair, so these four names have nowhere to land today, and the coupled decay plus the bias-corrected Adamax match what the torch recipe does. Will get this reviewed.

**Comment by danielhanchen** [comment]
Checked this against main and the four names still fall outside `SUPPORTED_MLX_OPTIMIZERS`, and the coupled decay plus the `_BiasCorrectedAdamax` subclass line up with torch.optim; will get this reviewed.
